File: Python_version/problem2.py
import matplotlib.pyplot as plt

k = 0.1
rod_length = 1
dx = 0.05
dt = 0.005
time_limit = 10

# adjusting
x_intervals = round(rod_length/dx)
dx = rod_length/x_intervals
x_node_count = x_intervals + 1

t_node_count = int(time_limit/dt) + 1
lamb = (k*dt)/(dx**2)

T = [[100 for i in range(x_node_count)] for j in range(t_node_count)]

# define initial condition
for t in range(t_node_count):
    T[t][0] = 300 #left boundary
    T[t][-1] = 300 #right boundary

# interior node counts
N = x_node_count - 2 

for t in range(t_node_count - 1):
    for i in range(N):
        T[t+1][i+1] = T[t][i+1] + lamb*(T[t][i+2] - 2*T[t][i+1] + T[t][i])

# what de helll kok jauh lebih gampang daripada implicit jir
for i in range(len(T)):
    print(T[i]) 

x = [i * dx for i in range(x_node_count)]

# Plot temperature profiles for each time step
for t in range(t_node_count):
    plt.plot(x, T[t], label=f't={round(t * dt, 2)}s')

plt.xlabel('Position')
plt.ylabel('Temperature')
plt.title('Temperature Profile Over Time (FTCS)')
plt.legend(loc='center left', bbox_to_anchor=(1, 0.5))
plt.grid(True)
plt.tight_layout()
plt.show()

File: Matlab_version/problem2.m
clc; 
clear;

% Program No. 2
L = 1; alpha = 0.1; Ti = 100; Ts = 300;
dx = 0.05; dt_values = [0.005, 0.01, 0.05];
x = 0:dx:L; nx = length(x);
t_final = 0.5;

for k = 1:length(dt_values)
    dt = dt_values(k);
    r = alpha * dt / dx^2;
    nt = round(t_final / dt);
    T = Ti * ones(1, nx);
    T(1) = Ts; T(end) = Ts;

    for n = 1:nt
        T_old = T;
        for i = 2:nx-1
            T(i) = T_old(i) + r * (T_old(i+1) - 2*T_old(i) + T_old(i-1));
        end
    end

    figure(k);
    plot(x, T, 'r-o'); grid on;
    title(["FTCS at \Delta t = ", num2str(dt), " hr"]);
    xlabel('x (ft)'); ylabel('Temperature (^\circF)');
end